File: frontend/src/types/preview.ts
import type { Orientation } from '../lib/pageDimensions'

export type PreviewDataResponse = {
  partset_id: string
  private_id: string
  title: string | null
  composer: string | null
  orientation: Orientation
  part_names: string[]
  combined_part_names: string[]
  part_segments: Record<string, number[]>
  segment_heights: number[]
  segment_widths: number[]
  segment_labels: string[]
  breaks: Record<string, number[]>
  spacings: Record<string, number>
  left_margin: number
  segment_urls: Record<string, string>
  images_ready: boolean
  images_warming: boolean
  image_progress: number
}

export type PartgenProgressResponse = {
  error: string | null
  status: string | null
  progress: number
  total_progress: number
  is_complete: boolean
  in_progress: boolean
}

export type PartsDataResponse = {
  partset_id: string
  private_id: string | null
  public_id: string
  mode: 'owner' | 'public'
  title: string | null
  composer: string | null
  publisher: string | null
  score_pdf_url: string | null
  parts: Array<{
    tag: string
    file_name: string
    letter_url: string
    a4_url: string
  }>
  parts_ready: boolean
  imslp_id: string | null
}

File: api/app/schemas/preview.py
from pydantic import BaseModel, Field


class PreviewDataResponse(BaseModel):
    partset_id: str
    private_id: str
    title: str | None = None
    composer: str | None = None
    part_names: list[str]
    combined_part_names: list[str]
    part_segments: dict[str, list[int]]
    segment_heights: list[float]
    segment_widths: list[float]
    segment_labels: list[str]
    breaks: dict[str, list[int]]
    spacings: dict[str, float]
    left_margin: int
    segment_urls: dict[str, str]
    images_ready: bool = True
    images_warming: bool = False
    image_progress: float = 100.0


class SaveLayoutRequest(BaseModel):
    breaks: dict[str, list[int]] = Field(default_factory=dict)
    spacings: dict[str, float] = Field(default_factory=dict)


class SaveLayoutResponse(BaseModel):
    status: str = "success"


class CombinePartsRequest(BaseModel):
    action: str
    tag: str


class CombinePartsResponse(BaseModel):
    status: str = "success"


class GeneratePartsResponse(BaseModel):
    status: str = "success"
    job_id: str | None = None


class PartgenProgressResponse(BaseModel):
    error: str | None = None
    status: str | None = None
    progress: float = 0.0
    total_progress: float = 0.0
    is_complete: bool = False


class PartDownloadItem(BaseModel):
    tag: str
    file_name: str
    letter_url: str
    a4_url: str


class PartsDataResponse(BaseModel):
    partset_id: str
    private_id: str | None = None
    public_id: str
    mode: str = "owner"
    title: str | None = None
    composer: str | None = None
    publisher: str | None = None
    score_pdf_url: str | None = None
    parts: list[PartDownloadItem]
    parts_ready: bool
